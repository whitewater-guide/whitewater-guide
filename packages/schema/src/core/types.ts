import { PageInfo } from '../__generated__/types';

/**
 * Relay-styles edge
 **/
export interface Edge<TNode> {
  node: TNode;
  cursor: CodegenCursor;
}

/**
 * Relay-styles connection
 **/
export interface RelayConnection<TNode> {
  edges: Array<Edge<TNode>>;
  pageInfo: PageInfo;
}
