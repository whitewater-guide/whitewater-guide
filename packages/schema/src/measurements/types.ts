export enum Unit {
  FLOW = 'flow',
  LEVEL = 'level',
}
