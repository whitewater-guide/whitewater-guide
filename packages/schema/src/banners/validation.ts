import { yupSchemas } from '@whitewater-guide/validation';
import type { ObjectSchema } from 'yup';
import { array, bool, mixed, number, object, string } from 'yup';

import type { BannerInput, BannerSourceInput } from '../__generated__/types';
import { BannerKind, BannerPlacement } from '../__generated__/types';

const BannerSourceSchema: ObjectSchema<BannerSourceInput> = object({
  kind: mixed<BannerKind>().oneOf(Object.values(BannerKind)).required(),
  url: string().required(),
}).noUnknown();

export const BannerInputSchema: ObjectSchema<BannerInput> = object({
  id: string().uuid().nullable(),
  slug: string().slug().required(),
  name: string().required(),
  priority: number().integer().required(),
  enabled: bool().required(),
  placement: mixed<BannerPlacement>()
    .oneOf(Object.values(BannerPlacement))
    .required(),
  source: BannerSourceSchema.clone(),
  link: string().url().https().nullable(),
  extras: object().nullable(),
  regions: array(yupSchemas.refInput().required()).required(),
  groups: array(yupSchemas.refInput().required()).required(),
})
  .strict(true)
  .noUnknown(true);
