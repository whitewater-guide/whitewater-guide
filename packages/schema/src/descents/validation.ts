import type { ObjectSchema } from 'yup';
import { bool, date, number, object, string } from 'yup';

import type { DescentInput, DescentLevelInput } from '../__generated__/types';

export const DescentLevelInputSchema: ObjectSchema<DescentLevelInput> = object({
  unit: string().max(20).optional().nullable(),
  value: number().required(),
})
  .strict(true)
  .noUnknown();

export const DescentInputSchema: ObjectSchema<DescentInput> = object({
  id: string().uuid().optional().nullable(),
  sectionId: string().uuid().required().nonNullable(),
  startedAt: date().required(),
  duration: number().integer().min(0).notRequired().nullable(),
  level: DescentLevelInputSchema.clone().notRequired().nullable(),
  comment: string().notRequired().nullable(),
  public: bool().notRequired().nullable(),
})
  .strict(true)
  .noUnknown();
