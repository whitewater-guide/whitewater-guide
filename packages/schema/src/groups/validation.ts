import type { ObjectSchema } from 'yup';
import { object, string } from 'yup';

import type { GroupInput } from '../__generated__/types';

const GROUP_SKU = /^group\.\w{3,}$/;

export const GroupInputSchema: ObjectSchema<GroupInput> = object({
  id: string().uuid().defined().nullable(),
  name: string().nonEmpty(),
  sku: string()
    .defined()
    .max(255)
    .matches(GROUP_SKU, 'yup:string.sku')
    .nullable(),
})
  .strict(true)
  .noUnknown();
