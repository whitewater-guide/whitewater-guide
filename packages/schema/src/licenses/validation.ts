import type { ObjectSchema } from 'yup';
import { object, string } from 'yup';

import type { LicenseInput } from '../__generated__/types';

export const LicenseInputSchema: ObjectSchema<LicenseInput> = object({
  url: string().url().nullable(),
  name: string().min(2).defined().nonNullable(),
  slug: string().slug().nullable(),
})
  .strict(true)
  .noUnknown();
