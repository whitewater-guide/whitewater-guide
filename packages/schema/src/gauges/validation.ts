import { yupSchemas } from '@whitewater-guide/validation';
import type { ObjectSchema } from 'yup';
import { mixed, object, string } from 'yup';

import type { GaugeInput } from '../__generated__/types';
import { PointInputSchema } from '../points';

export const GaugeInputSchema: ObjectSchema<GaugeInput> = object({
  id: string().uuid().nullable(),
  name: string().nonEmpty(),
  code: string().nonEmpty(),
  levelUnit: string().nullable(),
  flowUnit: string().nullable(),
  timezone: string().nullable(),
  location: PointInputSchema.clone().nullable(),
  requestParams: mixed().nullable(),
  url: string().url().nullable(),
  source: yupSchemas.refInput().required(),
})
  .strict(true)
  .noUnknown();
