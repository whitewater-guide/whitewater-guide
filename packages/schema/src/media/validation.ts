import type { ObjectSchema } from 'yup';
import { mixed, number, object, string, tuple } from 'yup';

import type { MediaInput } from '../__generated__/types';
import { MediaKind } from '../__generated__/types';
import { LicenseInputSchema } from '../licenses';

export const MediaInputSchema: ObjectSchema<MediaInput> = object({
  id: string().uuid().nullable(),
  description: string().nullable(),
  copyright: string().nullable(),
  license: LicenseInputSchema.clone().nullable(),
  url: string().nonEmpty(),
  kind: mixed<MediaKind>().defined().oneOf(Object.values(MediaKind)),
  resolution: tuple([
    number().required().integer().positive(),
    number().required().integer().positive(),
  ]).notRequired(),
  weight: number().integer().nullable(),
})
  .strict(true)
  .noUnknown();
