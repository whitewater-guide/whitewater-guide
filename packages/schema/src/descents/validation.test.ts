import { createSafeValidator } from '@whitewater-guide/validation';

import { DescentInput } from '../__generated__/types';
import { DescentInputSchema } from './validation';

const validator = createSafeValidator(DescentInputSchema);

type TestValue = [string, any];

const required: DescentInput = {
  sectionId: '879ca106-c264-11e8-a355-529269fb1459',
  startedAt: new Date('2021-06-20T15:48:46.458Z'),
};

const full: DescentInput = {
  ...required,
  id: '7ce3319e-c264-11e8-a355-529269fb1459',
  duration: 10,
  level: {
    value: 20,
    unit: 'cm',
  },
  comment: 'foo',
  public: true,
};

const correctValues: TestValue[] = [
  ['full value', full],
  ['required fields only', required],
  [
    'nulls',
    {
      ...required,
      id: null,
      duration: null,
      level: null,
      comment: null,
      public: null,
    },
  ],
  ['partial level', { ...required, level: { value: 300 } }],
];

const incorrectValues: TestValue[] = [
  ['bad uuid', { ...required, id: 'foo' }],
  ['bad section id', { ...required, sectionId: 'foo' }],
  ['bad date', { ...required, startedAt: 'foo' }],
  ['extra fields', { ...required, foo: 'bar' }],
];

it.each(correctValues)('should be valid for %s', (_, value) => {
  expect(validator(value)).toBeNull();
});

it.each(incorrectValues)('should be invalid for %s', (_, value) => {
  const error = validator(value);
  expect(error).not.toBeNull();
  expect(error).toMatchSnapshot();
});
