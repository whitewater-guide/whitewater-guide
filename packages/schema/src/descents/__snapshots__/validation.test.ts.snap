// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should be invalid for bad date 1`] = `
{
  "startedAt": {
    "key": "yup:mixed.notType",
    "options": {
      "type": "date",
    },
  },
}
`;

exports[`should be invalid for bad section id 1`] = `
{
  "sectionId": {
    "key": "yup:string.uuid",
    "options": {
      "path": "sectionId",
    },
  },
}
`;

exports[`should be invalid for bad uuid 1`] = `
{
  "id": {
    "key": "yup:string.uuid",
    "options": {
      "path": "id",
    },
  },
}
`;

exports[`should be invalid for extra fields 1`] = `
{
  "": {
    "key": "yup:object.noUnknown",
    "options": {
      "path": "this",
    },
  },
}
`;
